### ISSUE TITLE
Can you provide an average performance?

### ISSUE THREAD

**Issue by Jeonghoon4** [janghyuncho/PiCIE#3]
When I train the model, the performance is lower than the paper.
I think you also experimented with many models to get the best model.
I'm doing additional training now, and it would be very helpful if you could provide the results for previously trained models.

Can you provide the mean and standard deviation of the performance of the previously run models?
Can you provide performance for stuff only and thing only for models that do not use auxiliary clustering (+H)?

Thanks.

---------------------------------------------------------------------------------------
This is my conda environment for RTX 3090 GPU that supports CUDA>=11.0. 
python=3.7 -> 3.6
torch=1.2.0 -> 1.8.1
torchvision=0.4.0 -> 0.9.1
cudatoolkit=10.0 -> 11.0
faiss-gpu=1.6.3 -> 1.7.1

I trained 2 models using the train_picie.sh. 

**- stuff and things**
All ACC  : (1st) 39.3131 (± 16.0990) (2nd) 44.2331 (± 19.2532)
m IOU    : (1st) 9.0956                       (2nd) 10.9407

mean All ACC : 41.7731 ± 2.4600
best All ACC   : 44.2331 <---> 48.09 (paper), 49.99 (paper, +H)
mean mIOU    : 10.0182 ± 0.9226
best   mIOU    : 10.9407 <---> 13.84 (paper), 14.36 (paper, +H)

**- stuff**
All ACC  : (1st) 42.6251 (± 22.3882) (2nd) 48.9712 (± 27.5140)
m IOU    : (1st) 7.8613                       (2nd) 9.9796

mean All ACC : 45.7982 ± 3.1731
best All ACC   : 48.9712 <---> 74.56 (paper, +H)
mean mIOU    : 8.9205 ± 1.0592
best   mIOU    : 9.9796   <---> 17.32 (paper, +H)

**- thing**
All ACC  : (1st) 32.7959 (± 17.7133) (2nd) 47.6261 (± 26.1709)
m IOU    : (1st) 4.7463                       (2nd) 7.2779

mean All ACC : 40.2110 ± 7.4150
best All ACC   : 47.6261 <---> 69.39 (paper, +H)  
mean mIOU    : 6.0121 ± 1.2658
best   mIOU    : 7.2779   <---> 23.83 (paper, +H)

-------------------------------------------------------------------------------------------
Additionally, there are minor mistakes.

1) It is not working for evaluation of stuff only. (I modified it by using the thing only part.)
error part : [link](https://github.com/janghyuncho/PiCIE/blob/310e743386b929c230e37cb535dd1335ef9cf66e/data/coco_eval_dataset.py#L123)

2) 'is_first' option is not pre-defined. (first_batch?)
error part : [link](https://github.com/janghyuncho/PiCIE/blob/310e743386b929c230e37cb535dd1335ef9cf66e/train_picie.py#L304)
definition : [link](https://github.com/janghyuncho/PiCIE/blob/310e743386b929c230e37cb535dd1335ef9cf66e/commons.py#L50)

3) test bash code also require 'save_root'.

**Comment by janghyuncho** [comment]
Thank you for the feedback. 

1) We did find some difference in every run since (batch) clustering is inherently non-deterministic, but the deviation that you are seeing seems too much. Please let me investigate this closely and get back to you. There is a high chance that something went wrong during the refactoring - My apologies.

2) I am wondering how you computed the stuff-only and thing-only results? The code currently does not support this but the evaluation of a subset of classes is done by partitioning the confusion matrix (please refer to the [supplementary](https://openaccess.thecvf.com/content/CVPR2021/supplemental/Cho_PiCIE_Unsupervised_Semantic_CVPR_2021_supplemental.pdf), 2nd page). I added the evaluation code for the subsets [here](https://github.com/janghyuncho/PiCIE/blob/abd5e5b189ab793f13e8d1939133c1e2e1f054bd/commons.py#L230). 

3) Thank you for catching the mistakes. I will fix them as well. 

Again, thanks a lot for your feedback!

**Comment by Jeonghoon4** [comment]
I thought that the --thing and --stuff options in the previous test_picie.sh code reflect each subset, so I ran the code using only one of them. This is my mistake.

1) --stuff --thing     # for all data
2) --stuff                # for stuff only
3) --thing               # for thing only

I am very grateful for your quick feedback.

Thank you.

---------------------------------------------------------------------
This is my results using your fixed code.

**-- All results**

ACC  - All:
(mean) 41.461 ± 2.0565
(1) 39.3131
(2) 44.2331
(3) 40.8368

mIOU - All:
(mean) 10.3562 ± 0.8922
(1) 9.0956
(2) 10.9407
(3) 11.0325

---------------------------------------------------------------------
**-- Thing only**

ACC  - Thing: 
(mean) 68.7469 ± 14.3098
(1) 53.5455
(2) 64.7785
(3) 87.9168

mIOU - Thing: 
(mean) 13.9390 ± 2.6803
(1) 10.3783
(2) 14.5937
(3) 16.8451

---------------------------------------------------------------------
**-- Stuff only**

ACC  - Stuff:
(mean) 55.577 ± 7.1477
(1) 62.8217
(2) 58.0597
(3) 45.8496
       
mIOU - Stuff:
(mean) 19.7826 ± 0.9473
(1) 20.1302
(2) 20.7294
(3) 18.4884

**Comment by janghyuncho** [comment]
Hi, thank you for your patience. 

I have found an issue with the code from before and updated it just now. The main issues were that I was confused 1e-3 and 0.0001 for the learning rate and the clustering for each view was not synchronized. I fixed the code and confirmed that the result from the paper is reproducible with 5 runs with 47.6 +/- 1.2 Acc and 12.8 +/-1.1 mIOU.

**Comment by Jeonghoon4** [comment]
I could reproduce your work through fixed code.

This is the result of 2 runs.

**All**
ACC   : 47.7568 ± 0.5542
mIOU : 12.9189 ± 0.2382

**Thing**
ACC   : 67.1328 ± 1.4187
mIOU : 21.4697 ± 0.0332

**Stuff**
ACC   : 66.6689 ± 0.1188
mIOU : 19.0253 ± 0.1193

I also really appreciate your quick fix.

Thank you so much.
